src/editors/sharedComponents/InsertLinkModal/utils.js

```diff
@@ -0,0 +1,143 @@
+    const parentPath = block.type === blockTypes.section ? rootBlock.displayName : '';
+
+    addPathToBlocks(block, copyBlocks, blockRoot, null, parentPath);
+  });
+
+  return copyBlocks;
+};
+
+/**
+ * Retrieves a list of sections from the provided blocks object.
+ *
+ * @param {Object} blocks - The blocks object containing various block types.
+ * @returns {Array} An array of section (type: chapter) blocks extracted from the blocks object.
+ */
+export const getSectionsList = (blocks = {}) => {
+  const blocksList = Object.keys(blocks);
+  return blocksList.reduce((previousBlocks, blockKey) => {
+    const block = cloneDeep(blocks[blockKey]);
+    if (block.type === blockTypes.section) {
+      return [...previousBlocks, block];
+    }
+
+    return previousBlocks;
+  }, []);
+};
+
+/**
+ * Retrieves an array of child blocks based on the children list of a selected block.
+ *
+ * @param {Object} blockSelected - The selected block for which children are to be retrieved.
+ * @param {Object} blocks - The blocks object containing various block types.
+ * @returns {Array} An array of child blocks cloned from the blocks object.
+ */
+export const getChildrenFromList = (blockSelected, blocks) => {
+  if (blockSelected.children) {
+    return blockSelected.children.map((key) => cloneDeep(blocks[key]));
+  }
+  return [];
+};
+
+/**
+ * Filters blocks based on the provided searchText.
+ *
+ * @param {string} searchText - The text to filter blocks.
+ * @param {Object} blocks - The object containing blocks.
+ * @returns {Object} - Filtered blocks.
+ */
+export const filterBlocksByText = (searchText, blocks) => {
+  if (!searchText) {
+    return {};
+  }
+  const copyBlocks = cloneDeep(blocks);
+  return Object.keys(copyBlocks).reduce((result, key) => {
+    const item = copyBlocks[key];
+    if (item.path.toLowerCase().includes(searchText.toLowerCase())) {
+      result[key] = item;
+    }
+    return result;
+  }, {});
+};
+
+/**
+ * Formats a block path into title and subtitle.
+ *
+ * @param {string} path - The path to be formatted.
+ * @returns {Object} - Formatted block path with title and subtitle.
+ */
+export const formatBlockPath = (path) => {
+  if (!path) {
+    return {
+      title: '',
+      subTitle: '',
+    };
+  }
+  const pathSlitted = path.split(' / ');
+  let title = pathSlitted.pop();
+  const subTitle = pathSlitted.join(' / ');
+
+  if (!title.trim()) {
+    // If the last part is empty or contains only whitespace
+    title = pathSlitted.pop();
+  }
+  return {
+    title,
+    subTitle,
+  };
+};
+
+/**
+ * Validates a URL using a regular expression.
+ *
+ * @param {string} url - The URL to be validated.
+ * @returns {boolean} - True if the URL is valid, false otherwise.
+ */
+export const isValidURL = (url) => {
+  // Regular expression for a basic URL validation
+  const urlPattern = /^(https?|ftp):\/\/[^\s/$.?#].[^\s]*$/;
+
+  // Test the provided URL against the pattern
+  return urlPattern.test(url);
+};
```

Comment:
### Product
Are we OK with users *needing* to add `http://` or `https://` to their links? I could imagine users being frustrated when trying to add a link using just `example.com` doesn't work. 

That being said, it seems the default link flow in TinyMCE is also somewhat odd. When testing on https://www.tiny.cloud/docs/demo/basic-example/, I found that adding `example.com` as the URL leads to the link being treated as relative. Adding `www.example.com`, however, leads to a modal

![image](https://github.com/openedx/frontend-lib-content-components/assets/112954497/1e4352c7-7f08-4156-886a-eab7a50d83d2)

### Technical
I'm not sure if using regex for validation here is the right call. It should be possible to use the `URL()` constructor as a pretty much drop-in replacement https://developer.mozilla.org/en-US/docs/Web/API/URL/URL. We could also pass `type="url"` into the Paragon `Form.Control` and use `checkValidity()`. That being said, I'd be surprised if we weren't doing client-side URL validation anywhere else in the codebase. @arbrandes do you know if we have established best practices for this?

Reply:
For this case if the user is frustrated I understand that we are using this [mock up](https://www.figma.com/proto/cS42BX4OUsR8lOw6ppZHoz/Content-Libraries?type=design&node-id=1860-103852&t=6CUXwSKvDLmPMM61-0&scaling=min-zoom&page-id=1833%3A94107&starting-point-node-id=1860%3A103852&show-proto-sidebar=1) for this feature but there is not something like that, I can add it

For the URL() constructor I completely agree in use it here

Reply:
The linked mock specifically shows `www.example.com` (without `http://` or `https://`) as the example URL

![image](https://github.com/openedx/frontend-lib-content-components/assets/112954497/d8724159-26bf-4ad4-85c5-7b0666af5231)

In the current implementation that would fail this `isValidUrl` check.

Reply:
We are using a placeholder that indicates how the link should it be can I add the http:// or https:// to those urls but how could we know which one is?

Reply:
I will use http:// for those ones

Reply:
@felipemontoya @santiagosuarezedunext I'm hoping to get some product clarification on the "Link to URL" flow

PR comments may not be the best venue for this conversation, so I'm more than happy to find a time that works for a call if you'd rather discuss that way.

I mentioned some of it here https://github.com/openedx/frontend-lib-content-components/pull/455#discussion_r1490137186, but I was also hoping to hear some thoughts on https. 

Modern browsers use https by default to avoid making unnecessary insecure http requests. Unless a site does not support https, or a user explicitly enters `http://`, no insecure http requests are made when a user enters `example.com` into the address bar.

Many sites redirect from http to https, if you navigate to `http://github.com/`, for example, you will be redirected to `https://github.com/`. Other sites support https but do not redirect (see https://whynohttps.com/), and will simply serve the page using an insecure http connection instead (`http://example.com/` itself is an example of this).

There are also scenarios in which https is not available. If you clone this repo and serve it locally using `npm start`, it will be available at `http://localhost:8080/`, but not at `https://localhost:8080`

If we prepend `http://` to links created from urls entered without `http://` or `https://`, it is possible users will load the linked site insecurely. If we prepend `https://`, links to sites that are only served insecurely via http will be broken.

I guess I see 2 conflicting user stories here:
* As a course author (that cares about https), I want to be able to add https links without explicitly typing `https://`
* As a course author (that sometimes links to sites that don't support https), I want the links I add without explicitly typing `http://` to work.

Looking forward to hearing your thoughts!